### ISSUE TITLE
Batch Fetch hints ignored for simple "find by id" JPQL query because Eclipselink swaps out query internally

### ISSUE THREAD

**Issue by jnehlmeier** [eclipse-ee4j/eclipselink#862]
We have a very simple JPQL query
```sql
var jpql = SELECT o FROM SomeObject o WHERE o.id = :id
entityManager().createQuery(jpql, SomeObject.class)
   .setParameter("id", id)
   .setHint(QueryHints.BATCH_TYPE, BatchFetchType.IN)
   .setHint(QueryHints.BATCH, "o.relationA.relationB.relationC")
   .setHint(QueryHints.BATCH, "o.relationD.relationA.relationB.relationC")
   .getSingleResult();
```
and both batch fetches are ignored by EclipseLink and we see lots of small queries instead (n+1 problem).

However when changing the above query to 
```sql
var jpql = SELECT o FROM SomeObject o JOIN o.relationA ra WHERE o.id = :id AND ra IS NOT NULL
```
all batch fetches start to work. In our case the `relationA` is `@OneToOne` and `not null` so the modified query behaves the same as the original one.

Debugging revealed that the built `ReadObjectQuery` does have a correct `batchFetchPolicy` (in parent class `ObjectLevelReadQuery`)  assigned in both cases. However in the first case EclipseLink decides to throw away this query and replace it with another one which does not have a `batchFetchPolicy` assigned. Thus batch fetch does not work in the first case.

The code responsible is
https://github.com/eclipse-ee4j/eclipselink/blob/b7c997804f39187fb2907636d8403d694a318cfb/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/queries/DatabaseQuery.java#L858-L868
In line `866` the  query in `queryToExecute` (with `batchFetchPolicy`) will be replaced with `customQuery`(without `batchFetchPolicy`) which causes the issue to appear.

The code above calls `ObjectLevelReadQuery.checkForCustomQuery` which in turn calls `ReadObjectQuery.checkCustomQueryFlag`:
https://github.com/eclipse-ee4j/eclipselink/blob/b7c997804f39187fb2907636d8403d694a318cfb/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/queries/ReadObjectQuery.java#L319-L356
The large `if` statement in line `333` evaluates to `true` and finally the method returns `true` in line `348`. I am pretty sure this large `if` isn't complete and should evaluate to `false` in the example above. 

As can be seen the `if` checks for `JOIN` which is why the modified example fixes the issue. In that case the methods returns `NULL`, the original query will not be replaced by EclipseLink and batch fetch works as requested by application code.

**Comment by jnehlmeier** [comment]
I assume that
https://github.com/eclipse-ee4j/eclipselink/blob/b7c997804f39187fb2907636d8403d694a318cfb/foundation/org.eclipse.persistence.core/src/org/eclipse/persistence/queries/ReadObjectQuery.java#L331-L337
should contain a check wether or not any form of join fetch / batch fetch is requested for the query.

**Comment by xsunil** [comment]
I am also facing the same problem. Is there any work around for this.

**Comment by jnehlmeier** [comment]
I think you can add a nonsense where clause like AND 1=1 so that it is not a simple find by id query anymore

**Comment by patric-r** [comment]
Beside the actual issue - is it possible to specify multiple hints of the same type?
imho, eclipselink stores those hints in a Map therefore only storing the last one (see `org.eclipse.persistence.internal.jpa.QueryHintsHandler.Hint.apply(String, boolean, Object, DatabaseQuery, ClassLoader, AbstractSession)`)

**Comment by jnehlmeier** [comment]
@patric-r It seems the map you are referring to is only used to implement `Query.getHints()`, a JPA spec method. At least it is the only location the `QUERY_HINT_PROPERTIES` constant is used.

If you take a look at `return hint.apply(...)` every batch path will be parsed and added to the query: 
https://github.com/eclipse-ee4j/eclipselink/blob/7dfeab504e7b0042e6330315297bed0e6739cc95/jpa/org.eclipse.persistence.jpa/src/main/java/org/eclipse/persistence/internal/jpa/QueryHintsHandler.java#L1507

But yes, `Query.getHints()` would only return the last one defined via `Query.setHint()` for the same type of hint.

**Comment by patric-r** [comment]
You're right. I fell into the pitfall that EclipseLink added the queries but - possibly the same or similar case which we're discussing here - does not use them for some reason and then I confused this with the JPA methods.

What I really miss in EclipseLink is proper debug/trace logging for such aspects like hints (when/what are added, which are invalid, which one are ignored and why... etc.)
